hail/python/hailtop/batch/batch.py

```diff
@@ -149,6 +151,58 @@ def __init__(self,
 
         self._cancel_after_n_failures = cancel_after_n_failures
 
+        self._python_functions: Dict[int, Tuple[Callable, Optional[str]]] = {}
+
+    def _get_python_function(self, function_id: int) -> Callable:
+        function, _ = self._python_functions.get(function_id, (None, None))
+        if function is None:
+            raise ValueError(
+                f"No Python function registered with id: {function_id}."
+            )
+        return function
+
+    def _get_python_function_file_name(self, function_id: int) -> str:
+        _, file_name = self._python_functions.get(function_id, (None, None))
+        if file_name is None:
+            raise ValueError(
+                f"No file found for Python function registered with id: {function_id}."
+            )
+        return file_name
+
+    def _register_python_function(self, function: Callable) -> int:
+        function_id = id(function)
+        try:
+            self._get_python_function(function_id)
+        except ValueError:
+            self._python_functions[function_id] = (function, None)
+        return function_id
+
+    async def _serialize_python_to_input_file(
+        self, path: str, file_id: int, code: Any, dry_run: bool = False
+    ) -> str:
+        pipe = BytesIO()
+        dill.dump(code, pipe, recurse=True)
+        pipe.seek(0)
+
+        code_path = f"{path}/code{file_id}.p"
+
+        if not dry_run:
+            await self._fs.makedirs(os.path.dirname(code_path), exist_ok=True)
+            await self._fs.write(code_path, pipe.getvalue())
+
+        code_input_file = self.read_input(code_path)
+
+        return str(code_input_file)
```

Comment:
How come the cast this to a `str` here? The return of `read_input` should be fine to interpolate, and it would be great to preserve more expressive types where possible.

Reply:
Fixed to preserve the `InputResourceFile` until it is interpolated into the f-string for the command!